### PR TITLE
ddtrace/tracer: ensure access to trace tags is concurrency-safe

### LINKED ISSUE #1944: [BUG] 1.50.0: "fatal error: concurrent map read and map write"
**Version of dd-trace-go**
1.50.0 (upgrade from 1.48.0 is correlated with this issue)

**Describe what happened:**
fatal error! relevant go routine dump (with internal stacktrace bits elided):
```
"2023-04-26T22:05:11.509Z","fatal error: concurrent map read and map write"
"2023-04-26T22:05:11.509Z","goroutine 101135 [running]:"
"2023-04-26T22:05:11.509Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*propagatorW3c).injectTextMap(0xc03a4f7ad8, {0x45fcad0?, 0xc03ecfc2a0?}, {0x7f8aade2fa60, 0xc03c27d650})"
"2023-04-26T22:05:11.509Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/textmap.go:674 +0x312"
"2023-04-26T22:05:11.509Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*propagatorW3c).Inject(0xc03a4f7ac8?, {0x45fcad0, 0xc03ecfc2a0}, {0x38f1b40?, 0xc03c27d650?})"
"2023-04-26T22:05:11.510Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/textmap.go:630 +0x89"
"2023-04-26T22:05:11.510Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*chainedPropagator).Inject(0xc00007ec00?, {0x45fcad0, 0xc03ecfc2a0}, {0x38f1b40, 0xc03c27d650})"
"2023-04-26T22:05:11.510Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/textmap.go:246 +0x83"
"2023-04-26T22:05:11.510Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*tracer).Inject(0xc035b934a0?, {0x45fcad0?, 0xc03ecfc2a0?}, {0x38f1b40?, 0xc03c27d650?})"
"2023-04-26T22:05:11.510Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/tracer.go:595 +0x3f"
"2023-04-26T22:05:11.510Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.Inject({0x45fcad0?, 0xc03ecfc2a0?}, {0x38f1b40?, 0xc03c27d650?})"
"2023-04-26T22:05:11.511Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/tracer.go:180 +0x96"

"2023-04-26T22:05:11.646Z","goroutine 2990 [select]:"
"2023-04-26T22:05:11.646Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*concentrator).runFlusher(...)"
"2023-04-26T22:05:11.646Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/stats.go:106"
"2023-04-26T22:05:11.646Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*concentrator).Start.func1()"
"2023-04-26T22:05:11.646Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/stats.go:94 +0x136"
"2023-04-26T22:05:11.646Z","created by gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*concentrator).Start"
"2023-04-26T22:05:11.646Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/stats.go:90 +0xb8"

"2023-04-26T22:05:11.695Z","goroutine 2989 [select]:"
"2023-04-26T22:05:11.696Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*tracer).reportHealthMetrics(0xc014a33900, 0x0?)"
"2023-04-26T22:05:11.696Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/metrics.go:101 +0x1bf"
"2023-04-26T22:05:11.696Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.newTracer.func3()"
"2023-04-26T22:05:11.696Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/tracer.go:277 +0x65"
"2023-04-26T22:05:11.696Z","created by gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.newTracer"
"2023-04-26T22:05:11.696Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/tracer.go:275 +0x196"

"2023-04-26T22:05:11.753Z","goroutine 2991 [select, 8 minutes]:"
"2023-04-26T22:05:11.753Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*concentrator).runIngester(0xc00db600c0)"
"2023-04-26T22:05:11.753Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/stats.go:127 +0xae"
"2023-04-26T22:05:11.753Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*concentrator).Start.func2()"
"2023-04-26T22:05:11.753Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/stats.go:99 +0x5a"
"2023-04-26T22:05:11.754Z","created by gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*concentrator).Start"
"2023-04-26T22:05:11.754Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/stats.go:97 +0x105"

"2023-04-26T22:05:11.829Z","goroutine 2988 [select]:"
"2023-04-26T22:05:11.829Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*tracer).worker(0xc014a33900, 0xc00dbf4660)"
"2023-04-26T22:05:11.830Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/tracer.go:310 +0xdc"
"2023-04-26T22:05:11.830Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.newTracer.func2()"
"2023-04-26T22:05:11.830Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/tracer.go:272 +0xba"
"2023-04-26T22:05:11.830Z","created by gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.newTracer"
"2023-04-26T22:05:11.830Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/tracer.go:264 +0x145"

"2023-04-26T22:05:11.891Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.(*tracer).reportRuntimeMetrics(0xc014a33900, 0x2540be400?)"
"2023-04-26T22:05:11.891Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/metrics.go:44 +0x1b3"
"2023-04-26T22:05:11.891Z","gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.newTracer.func1()"
"2023-04-26T22:05:11.891Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/tracer.go:260 +0x68"
"2023-04-26T22:05:11.891Z","created by gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.newTracer"
"2023-04-26T22:05:11.891Z",".../vendor/gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer/tracer.go:258 +0xe5"
```

**Describe what you expected:**
no fatal errors

**Steps to reproduce the issue:**
I'm working on this but I figured the go routine dump + general awareness might be helpful to start.

**Additional environment details (Version of Go, Operating System, etc.):**
running go 1.20.3

### PR DESCRIPTION
**Pull request by knusbaum**
### What does this PR do?

This commit adds a few methods to trace to allow safe access to the tags and propagatingTags members of trace to the marshaling code.

Fixes #1944 

### Motivation

Spancontext marshaling was accessing tracer internal structures without a lock, resulting in a data race and panic.
### Reviewer's Checklist

- [ ] Changed code has unit tests for its functionality.
- [ ] If this interacts with the agent in a new way, a system test has been added.

### CODE REVIEW

File: ddtrace/tracer/textmap_test.go [hunk 1]
```diff
@@ -1964,3 +1965,22 @@ func FuzzExtractTraceID128(f *testing.F) {
...
 	})
 }
+
+// Regression test for https://github.com/DataDog/dd-trace-go/issues/1944
+func TestPropagatingTagsConcurrency(t *testing.T) {
+	trc := newTracer()
+	defer trc.Stop()
+
+	var wg sync.WaitGroup
+	for i := 0; i < 10000; i++ {
```
Comment (ajgajg1134) [inline ddtrace/tracer/textmap_test.go]: nit 
```suggestion
	for i := 0; i < 10_000; i++ {
```

File: ddtrace/tracer/textmap_test.go [hunk 2]
```diff
@@ -1964,3 +1965,22 @@ func FuzzExtractTraceID128(f *testing.F) {
...
 	})
 }
+
+// Regression test for https://github.com/DataDog/dd-trace-go/issues/1944
+func TestPropagatingTagsConcurrency(t *testing.T) {
+	trc := newTracer()
+	defer trc.Stop()
+
+	var wg sync.WaitGroup
+	for i := 0; i < 1000; i++ {
```
Comment (ajgajg1134) [inline ddtrace/tracer/textmap_test.go:1975]: nit
```suggestion
	for i := 0; i < 1_000; i++ {
```

File: ddtrace/tracer/span.go [hunk 3]
```diff
@@ -221,7 +221,7 @@ func (s *span) SetUser(id string, opts ...UserMonitoringOption) {
 	} else {
 		// Unset the propagated user ID so that a propagated user ID coming from upstream won't be propagated anymore.
 		trace.unsetPropagatingTag(keyPropagatedUserID)
-		if _, ok := trace.propagatingTags[keyPropagatedUserID]; ok {
+		if trace.hasPropagatingTag(keyPropagatedUserID) {
```
Comment (ajgajg1134) [inline ddtrace/tracer/span.go:224]: Is it just me or does this `if` never happen since we just unset the tag on the line above?

**Review by ajgajg1134 — CHANGES_REQUESTED** [review]
Overall looks good, just a few small nits

File: ddtrace/tracer/span.go (hunk 3 above)
Comment (knusbaum) [reply ddtrace/tracer/span.go:224]: indeed it does.

**Review by ajgajg1134 — APPROVED** [review]

**Pull request merged**